### DISCUSSION TITLE
Authentification is not working in Cloud Run

### DISCUSSION THREAD

**Discussion by NastuzziSamy** [spatie/laravel-google-cloud-storage#64] [Q&A]
Hi, thank you for the package.

I managed to make it work in Cloud Build but in cloud run, with the same custom service (with the required permissions), I am stucked with the following error:

```
Client error: `POST https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/[service]@[project].iam.gserviceaccount.com:signBlob?alt=json` resulted in a `403 Forbidden` response:
{
  "error": {
    "code": 403,
    "message": "Permission 'iam.serviceAccounts.implicitDelegation' denied on resource (or it may not exist).",
    "status": "PERMISSION_DENIED",
    "details": [
      {
        "@type": "type.googleapis.co (truncated...)
```

**Comment by thehands79** [comment]
I have the same with a Laravel app on Google App Engine standard. I use [Application Default Credentials](https://cloud.google.com/docs/authentication/provide-credentials-adc) to authenticate from my local machine, which works perfectly. When it runs on app engine using the App Engine default service account and the [default bucket](https://cloud.google.com/appengine/docs/standard/using-cloud-storage?tab=php#default_bucket) that app engine created for me, I get the 403. My env vars and error:

```
GOOGLE_CLOUD_PROJECT_ID=myproject
GOOGLE_CLOUD_STORAGE_BUCKET=myproject.appspot.com
GOOGLE_CLOUD_STORAGE_PATH_PREFIX=photos

2023-12-05 16:05:29 default[xyz]   [2023-12-05 16:05:29] dev.ERROR: Client error: `POST https://iamcredentials.googleapis.com/v1/projects/-/serviceAccounts/myproject@appspot.gserviceaccount.com:signBlob?alt=json` resulted in a `403 Forbidden` response:
2023-12-05 16:05:29 default[xyz]  {
2023-12-05 16:05:29 default[xyz]    "error": {
2023-12-05 16:05:29 default[xyz]      "code": 403,
2023-12-05 16:05:29 default[xyz]      "message": "Permission 'iam.serviceAccounts.implicitDelegation' denied on resource (or it may not exist).",
2023-12-05 16:05:29 default[xyz]      "status": "PERMISSION_DENIED",
2023-12-05 16:05:29 default[xyz]      "details": [
2023-12-05 16:05:29 default[xyz]        {
2023-12-05 16:05:29 default[xyz]          "@type": "type.googleapis.co (truncated...)
```

I can browse the bucket in GCP console and it definitely exists. According to the docs the default service account [has all the necessary IAM permissions](https://cloud.google.com/appengine/docs/standard/using-cloud-storage?tab=php#set_bucket_permissions).

**Comment by izzudin96** [comment]
@NastuzziSamy @thehands79 did you manage to fix the issue?

**Reply by thehands79** [reply]
If I recall correctly, the workaround was bundling the JSON key with the app, which is poor security practice but couldn't make it work otherwise.

**Comment by zaxbux** [comment]
The readme example has `key_file` set to an empty array, which means empty credentials are passed to the StorageClient. Set to `null` or remove it.

`filesystems.php` config:

```php
'gcs' => [
    'driver' => 'gcs',
    // ...
    'key_file' => null, // remove or set to `null` to use the service account bound to Cloud Run / Compute Engine
    // ...
],
```
